adapters/insticator/insticator.go

```diff
@@ -0,0 +1,526 @@
+		}
+
+		if len(request.Device.IPv6) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+			headers.Add("IP", request.Device.IP)
+		}
+	}
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}, nil
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for i := range seatBid.Bid {
+			bid := &seatBid.Bid[i]
+			bidType := getMediaTypeForBid(bid)
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: bidType,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+	return bidResponse, nil
+}
+
+func makeImps(imp openrtb2.Imp) (openrtb2.Imp, error) {
+	if imp.Banner == nil && imp.Video == nil {
+		return openrtb2.Imp{}, &errortypes.BadInput{
+			Message: fmt.Sprintf("Imp ID %s must have at least one of [Banner, Video] defined", imp.ID),
+		}
+	}
+
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return openrtb2.Imp{}, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var insticatorExt openrtb_ext.ExtImpInsticator
+	if err := json.Unmarshal(bidderExt.Bidder, &insticatorExt); err != nil {
+		return openrtb2.Imp{}, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var impExt Ext
+	impExt.Insticator.AdUnitId = insticatorExt.AdUnitId
+	impExt.Insticator.PublisherId = insticatorExt.PublisherId
+
+	impExtJSON, err := json.Marshal(impExt)
+	if err != nil {
+		return openrtb2.Imp{}, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	imp.Ext = impExtJSON
+
+	// Validate Video if it exists
+	if imp.Video != nil {
+		videoCopy, err := validateVideoParams(imp.Video)
+
+		imp.Video = videoCopy
+
+		if err != nil {
+			return openrtb2.Imp{}, &errortypes.BadInput{
+				Message: err.Error(),
+			}
+		}
```

Comment:
- should remove line space on line 381 and line 384
- error check should be done before assigning value to `imp.Video`
- create copy of imp in `makeImps` and then make changes in imp copy

```
videoCopy, err := validateVideoParams(imp.Video)
if err != nil {
	return openrtb2.Imp{}, &errortypes.BadInput{
		Message: err.Error(),
	}
}
imp.Video = videoCopy
```